### ISSUE TITLE
Unable to build DKMS module with 4.8 (rc5)

### ISSUE THREAD

**Issue by nathan-osman** [corngood/archlinux-amdgpu#16]
I just built and installed kernel 4.8.0-rc5-g3eab887 (x86_64) using the `linux-git` package on AUR. I tried to build the `archlinux-amdgpu` package but run into the following errors when building the DKMS modules:

```
DKMS make.log for amdgpu-pro-16.30.3-315407 for kernel 4.8.0-rc5-g3eab887 (x86_64)
Thu Sep  8 20:21:48 PDT 2016
make: Entering directory '/usr/lib/modules/4.8.0-rc5-g3eab887/build'
  LD      /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/built-in.o
  LD      /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/built-in.o
  CC [M]  /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_drv.o
  CC [M]  /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_device.o
  CC [M]  /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_kms.o
  CC [M]  /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_atombios.o
  CC [M]  /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/atombios_crtc.o
  CC [M]  /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_connectors.o
  CC [M]  /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/atom.o
  CC [M]  /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_fence.o
  CC [M]  /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_ttm.o
/var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_ttm.c: In function ‘amdgpu_move_blit’:
/var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_ttm.c:291:18: error: incompatible type for argument 4 of ‘ttm_bo_move_accel_cleanup’
           evict, no_wait_gpu, new_mem);
                  ^~~~~~~~~~~
In file included from /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/../backport/include/kcl/kcl_ttm.h:6:0,
                 from /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/../backport/backport.h:9,
                 from <command-line>:0:
./include/drm/ttm/ttm_bo_driver.h:1032:12: note: expected ‘struct ttm_mem_reg *’ but argument is of type ‘bool {aka _Bool}’
 extern int ttm_bo_move_accel_cleanup(struct ttm_buffer_object *bo,
            ^~~~~~~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_ttm.c:290:6: error: too many arguments to function ‘ttm_bo_move_accel_cleanup’
  r = ttm_bo_move_accel_cleanup(bo, fence,
      ^~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/../backport/include/kcl/kcl_ttm.h:6:0,
                 from /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/../backport/backport.h:9,
                 from <command-line>:0:
./include/drm/ttm/ttm_bo_driver.h:1032:12: note: declared here
 extern int ttm_bo_move_accel_cleanup(struct ttm_buffer_object *bo,
            ^~~~~~~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_ttm.c: In function ‘amdgpu_move_vram_ram’:
/var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_ttm.c:337:6: error: too few arguments to function ‘ttm_bo_move_ttm’
  r = ttm_bo_move_ttm(bo, true, no_wait_gpu, new_mem);
      ^~~~~~~~~~~~~~~
In file included from /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/../backport/include/kcl/kcl_ttm.h:6:0,
                 from /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/../backport/backport.h:9,
                 from <command-line>:0:
./include/drm/ttm/ttm_bo_driver.h:979:12: note: declared here
 extern int ttm_bo_move_ttm(struct ttm_buffer_object *bo,
            ^~~~~~~~~~~~~~~
/var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_ttm.c: In function ‘amdgpu_move_ram_vram’:
/var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_ttm.c:370:2: error: the address of ‘tmp_mem’ will always evaluate as ‘true’ [-Werror=address]
  r = ttm_bo_move_ttm(bo, true, no_wait_gpu, &tmp_mem);
  ^
/var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_ttm.c:370:6: error: too few arguments to function ‘ttm_bo_move_ttm’
  r = ttm_bo_move_ttm(bo, true, no_wait_gpu, &tmp_mem);
      ^~~~~~~~~~~~~~~
In file included from /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/../backport/include/kcl/kcl_ttm.h:6:0,
                 from /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/../backport/backport.h:9,
                 from <command-line>:0:
./include/drm/ttm/ttm_bo_driver.h:979:12: note: declared here
 extern int ttm_bo_move_ttm(struct ttm_buffer_object *bo,
            ^~~~~~~~~~~~~~~
/var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_ttm.c: In function ‘amdgpu_bo_move’:
/var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_ttm.c:432:7: error: too few arguments to function ‘ttm_bo_move_memcpy’
   r = ttm_bo_move_memcpy(bo, evict, no_wait_gpu, new_mem);
       ^~~~~~~~~~~~~~~~~~
In file included from /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/../backport/include/kcl/kcl_ttm.h:6:0,
                 from /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/../backport/backport.h:9,
                 from <command-line>:0:
./include/drm/ttm/ttm_bo_driver.h:1002:12: note: declared here
 extern int ttm_bo_move_memcpy(struct ttm_buffer_object *bo,
            ^~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [scripts/Makefile.build:290: /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu/amdgpu_ttm.o] Error 1
make[1]: *** [scripts/Makefile.build:440: /var/lib/dkms/amdgpu-pro-16.30.3/315407/build/amd/amdgpu] Error 2
make: *** [Makefile:1471: _module_/var/lib/dkms/amdgpu-pro-16.30.3/315407/build] Error 2
make: Leaving directory '/usr/lib/modules/4.8.0-rc5-g3eab887/build'

```

**Comment by corngood** [comment]
Damn, looks like more patches will be needed for 4.8, and it doesn't look like it's been done at https://github.com/Lucretia/vulkan-overlay (gentoo) yet.

I don't have much time to work on this at the moment, so I'd personally rather wait for 4.8 to be released, but if anyone wants to do a PR that doesn't break 4.7, I'll take it.

**Comment by nathan-osman** [comment]
Some of the errors do look like easy fixes. I may take a stab at this.

**Comment by nathan-osman** [comment]
The necessary changes are as follows (I haven't yet figured out how to properly apply a patch to the module source code):

- `amd/amdgpu/amdgpu_ttm.c:291` - remove `no_wait_gpu` parameter
- `amd/amdgpu/amdgpu_ttm.c:337`, `370`, &amp; `432` - insert `false` after second parameter
- `amd/amdgpu/amdgpu_display.c:269` - change line to `r = drm_crtc_vblank_get(crtc);`
- `amd/amdgpu/amdgpu_display.c:294` - change line to `r = drm_crtc_vblank_put(crtc);`
- `amd/amdgpu/amdgpu_sa.c:430` - change `%d` to `%lld`
- `amd/amdgpu/dce_v8_0.c:2577`, `amd/amdgpu/dce_v10_0.c:2670`, &amp; `amd/amdgpu/dce_v11_0.c:2680` - change return type to `int`
- `amd/amdgpu/dce_v8_0.c:2578`, `amd/amdgpu/dce_v10_0.c:2671`, &amp; `amd/amdgpu/dce_v11_0.c:2681` - remove `start` parameter
- `amd/amdgpu/dce_v8_0.c:2581`, `amd/amdgpu/dce_v10_0.c:2674`, &amp; `amd/amdgpu/dce_v11_0.c:2684` - change line to `int i;`
- `amd/amdgpu/dce_v8_0.c:2584`, `amd/amdgpu/dce_v10_0.c:2677`, &amp; `amd/amdgpu/dce_v11_0.c:2687` - change line to `for (i = 0; i < size; i++) {`
- `amd/amdgpu/dce_v8_0.c:2589`, `amd/amdgpu/dce_v10_0.c:2682`, &amp; `amd/amdgpu/dce_v11_0.c:2692` - add `return 0;`
- `amd/amdgpu/dce_v8_0.c:3375`, `amd/amdgpu/dce_v10_0.c:3371`, &amp; `amd/amdgpu/dce_v11_0.c:3431` - change line to `drm_crtc_send_vblank_event(&amdgpu_crtc->base, works->event);`
- `amd/amdgpu/dce_v8_0.c:3379`, `amd/amdgpu/dce_v10_0.c:3375`, &amp; `amd/amdgpu/dce_v11_0.c:3435` - change line to `drm_crtc_vblank_put(&amdgpu_crtc->base);`
- `amd/amdgpu/amdgpu_acp.c:424-446` - remove lines
- `amd/dal/amdgpu_dm/amdgpu_dm_types.c:982` - change return type to `int`
- `amd/dal/amdgpu_dm/amdgpu_dm_types.c:987` - remove line
- `amd/dal/amdgpu_dm/amdgpu_dm_types.c:994` - add `return 0;`
- `amd/dal/amdgpu_dm/amdgpu_dm_types.c:2221` - change function parameters to `state, true`
- `amd/dal/amdgpu_dm/amdgpu_dm.c:199` - change function to `drm_crtc_send_vblank_event`
- `amd/dal/amdgpu_dm/amdgpu_dm.c:200` - change line to `&amdgpu_crtc->base,`
- `amd/dal/amdgpu_dm/amdgpu_dm.c:201` - remove line
- `amd/backport/kcl_ttm.c:111` - change to `bo->moving = NULL;`

**Comment by nathan-osman** [comment]
If someone can show me how to properly create a patch, I can convert my list of changes into something that can be applied more easily :stuck_out_tongue:

**Comment by nathan-osman** [comment]
In the meantime, I have converted this into a diff so that it would be slightly easier to apply:

[patch.txt](https://github.com/corngood/archlinux-amdgpu/files/463339/patch.txt)

_(Apparently GitHub doesn't like files with `.diff` extension.)_

**Comment by corngood** [comment]
Thanks for the patch.  A patch in that format is suitable to be dropped in the root of the project (e.g. as `0011-linux-4.8-fixes.patch`.  Then you just need to run `build.sh` to generate the PKGBUILD.

I think most of it will need to be wrapped in `#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 8, 0)`.  Other than that, the only thing that concerns me is the removal of the contents of `acp_resume`.  Why did you need to do that?

When you've tested it could you let me know how it went, and which GPU you used?

Cheers

**Comment by NogradThGin** [comment]
I will work on a patch for 4.8 in this week i'm on 4.8 too and succeded to install DKMS for it (all working fine) using Linux 4.8 (Git version) + BFS/BFQ Patch and GCC CPU Optimizations Patch (kernel part).

The only thing is amdgpu folder  is over 1000 lines modified (such a pain for creating a patch for all thing :dart: )

I will do it anyway (first a "primitive" patch for make it work) (and second a full patch containing all modification on 4.8)

**Comment by corngood** [comment]
Do you think it's worth looking at the steamos version (#17)?  It's possible the driver in (16.40) is compatible with newer kernels.  It's kind of a long shot since I think SteamOS is on 4.1, but I haven't looked at it at all.

**Comment by pattop** [comment]
The removal of the contents of acp_resume is consistent with commit 39dd0f234fc37da071dadbd9b49fe800d62139b4 in linux.

I successfully built and ran a driver with these changes on an RX 480 under kernel 4.8.6 which worked fine.

Unfortunately it still doesn't help me run my Dell UP3214Q at 60Hz using MST :(

**Comment by corngood** [comment]
This will be fixed in 16.60, see #28
